### PR TITLE
When profile_memory, also export and save the snapshot.pickle for lora_finetune_single_device.py

### PR DESCRIPTION
**Pull request by janeyx99**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* enable the snapshot.pickle to be exported when profile_memory is True.
* the above required adding attributes to self for recipes/lora_finetune_single_device.py
* needing to export the snapshot.pickle _does_ mean that it takes a little longer to export.

Previously, you only had the html from the profiler, which is not interactive:
image

Now, you get that in addition to:
image

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](../CONTRIBUTING.md) for some guidance on contributing.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](../tests/torchtune) for any new functionality
- [ ] update [docstrings](../docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [ ] I have added an example to docs or docstrings;

### CODE REVIEW

**Review by ebsmothers — APPROVED** [review]
Thanks for adding this! Can you fix the linter complaints? After that this should be good to go

File: recipes/lora_finetune_single_device.py [hunk 1]
```diff
@@ -579,6 +584,9 @@ def train(self) -> None:
                     ):
                         break
```
Comment (felipemello1) [inline recipes/lora_finetune_single_device.py:586]:
```suggestion
                    # Memory profiling
                    if curr_epoch == 0 and self.profiler_profile_memory and idx == (self.profiler_wait_steps + self.profiler_warmup_steps):
                        torch.cuda.memory._record_memory_history()
```

File: recipes/lora_finetune_single_device.py [hunk 2]
```diff
@@ -579,6 +584,9 @@ def train(self) -> None:
...
 
+                    if curr_epoch == 0 and self.profiler_profile_memory and idx == self.profiler_wait_steps + self.profiler_warmup_steps:
+                        torch.cuda.memory._record_memory_history()
```
Comment (felipemello1) [inline recipes/lora_finetune_single_device.py:588]: The issue is if in the future we support other backend, like intel. It makes me think that this should be a function, instead of being hardcoded as "torch.cuda"

cc: @ebsmothers

File: recipes/lora_finetune_single_device.py [hunk 3]
```diff
@@ -626,6 +634,9 @@ def train(self) -> None:
                         num_tokens = 0
                         t0 = time.perf_counter()
 
+                    if curr_epoch == 0 and self.profiler_profile_memory and idx == self.profiler_wait_steps + self.profiler_warmup_steps + self.profiler_active_steps:
```
Comment (felipemello1) [inline recipes/lora_finetune_single_device.py:637]:
```suggestion
                    # Stop memory profiling
                    if curr_epoch == 0 and self.profiler_profile_memory and idx == self.profiler_wait_steps + self.profiler_warmup_steps + self.profiler_active_steps:
```

**Review by felipemello1 — COMMENTED** [review]
@ebsmothers, wdyt? Were were missing the memory snapshot. My only concern is hardcoding torch.cuda if we plan to support xpu

**Comment by felipemello1** [comment]
should we just delete the HTML to save time, since it takes a while to save? Or is it useful?

File: recipes/lora_finetune_single_device.py (hunk 2 above)
Comment (ebsmothers) [reply recipes/lora_finetune_single_device.py:588]: > The issue is if in the future we support other backend, like intel. It makes me think that this should be a function, instead of being hardcoded as "torch.cuda"
> 
> cc: @ebsmothers

I think it's OK, idk that memory snapshot would be supported on XPU anyways. We can do a check on device + profile_memory combo just to be safe though

File: recipes/lora_finetune_single_device.py [hunk 4]
```diff
@@ -328,6 +328,11 @@ def _setup_profiler(
 
         log.info(f" Profiler config after instantiation: {profiler_cfg}")
 
+        self.profiler_wait_steps = profiler_cfg["wait_steps"]
```
Comment (ebsmothers) [inline recipes/lora_finetune_single_device.py:331]: This seems to be throwing an error in our recipe tests, can you just quickly inspect `profiler_cfg` after running e.g.

```
pytest tests/recipes/test_lora_finetune_single_device.py -m integration_test -k 'test_loss'
```

I thought these fields would be defined based on [this](https://github.com/pytorch/torchtune/blob/2f8ed7a34457eaddce59a58e8c37dd51d6ba65f8/torchtune/utils/_profiler.py#L370), but seems like something weird is happening here.

Comment (janeyx99) [reply recipes/lora_finetune_single_device.py:331]: Ah, it's because profiling isn't enabled for tests, so it returns here https://github.com/pytorch/torchtune/blob/2f8ed7a34457eaddce59a58e8c37dd51d6ba65f8/torchtune/utils/_profiler.py#L262

**Comment by janeyx99** [comment]
> should we just delete the HTML to save time, since it takes a while to save? Or is it useful?

Writing the html is very light so it doesn't make a tangible difference, I'm not sure if people are using it today so it'd be bad to just remove it if there were actual users.
